### PR TITLE
Move exemplar types to non-internal package

### PR DESCRIPTION
**Pull request by dashpole**
Part of https://github.com/open-telemetry/opentelemetry-go/issues/5249

This makes all existing types designed to implement the public Exemplar API public by moving most of `internal/exemplar` to `exemplar`.  The only types that are not being made public are `exemplar.Drop`, and `exemplar.FilteredReservoir`.  Those types are moved to `internal/aggregate`, and are renamed to `DropReservoir` and `FilteredExemplarReservoir`.

The following types are made public:

* `exemplar.Exemplar`
* `exemplar.Filter`
* `exemplar.SampledFilter`
* `exemplar.AlwaysOnFilter`
* `exemplar.HistogramReservoir`
* `exemplar.FixedSizeReservoir`
* `exemplar.Reservoir`
* `exemplar.Value`
* `exemplar.ValueType`

### CODE REVIEW

**Comment by dashpole** [comment]
The README link doesn't work yet because the package isn't released.  I assume I should just leave it there anyways?

File: sdk/metric/exemplar/README.md
Comment (dashpole) [inline sdk/metric/exemplar/README.md:1]: TODO: update

Comment (dashpole) [inline sdk/metric/exemplar/README.md:1]:
```suggestion
# Metric SDK Exemplars
```

**Comment by dashpole** [comment]
TODO: list all of the types being made public in the description

**Comment by dashpole** [comment]
@MrAlias rebased

**Review by XSAM — APPROVED** [review]
👍

**Comment by MrAlias** [comment]
Should we rename FixedSize and Histogram to FixedSizeReservoir and HistogramReservoir?

**Comment by XSAM** [comment]
> Should we rename FixedSize and Histogram to FixedSizeReservoir and HistogramReservoir?

I vote for this.

**Comment by dashpole** [comment]
Renamed.  I also took the liberty to rename the files to fixed_size_reservoir.go and histogram_reservoir.go while we are renaming things...

**Comment by dashpole** [comment]
Possibly a follow-up question, but should we name these `NewFixedSizeReservoir` and `NewHistogramReservoir`, and export concrete types `FixedSizeReservoir` and `HistogramReservoir`?  I can't think of why we would need methods on those other than implementing the Reservoir interface, but it is a generally good practice to accept interfaces and return concrete types.

**Comment by XSAM** [comment]
> should we name these NewFixedSizeReservoir and NewHistogramReservoir

Looks good.

> export concrete types FixedSizeReservoir and HistogramReservoir

I don't have opinions on this.

**Comment by MrAlias** [comment]
> Possibly a follow-up question, but should we name these `NewFixedSizeReservoir` and `NewHistogramReservoir`, and export concrete types `FixedSizeReservoir` and `HistogramReservoir`? I can't think of why we would need methods on those other than implementing the Reservoir interface, but it is a generally good practice to accept interfaces and return concrete types.

Both of these ideas sounds good to me, :+1:

**Review by XSAM — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Comment by dashpole** [comment]
@MrAlias i'm waiting for you review before merging.  Let me know if I should merge without your approval.

File: CHANGELOG.md
Comment (MrAlias) [inline CHANGELOG.md:17]: The style of this document is wrapped on sentence. Can we stick wit this style? Maybe add the types in a bulleted list if they need to be collapsed.

File: sdk/metric/exemplar/value.go
Comment (MrAlias) [inline sdk/metric/exemplar/value.go:1]: Is `NewValue` extensible? If we want to add support for `uint64` values, won`t `NewValue[N int64 | float64 | uint64](value N) Value` be a different function signature?

Should we have typed creation functions similar to `attribute`s?

File: sdk/metric/exemplar/fixed_size_reservoir.go
Comment (MrAlias) [inline sdk/metric/exemplar/fixed_size_reservoir.go:48]:
```suggestion
func newFixedSizeReservoir(s *storage) *FixedSizeReservoir {
```

Right?

File: sdk/metric/exemplar/value.go
Comment (MrAlias) [reply sdk/metric/exemplar/value.go:1]: This is not a concern. This should be forward-compatible in the ways we need it to be. This would also be an issue for the `metricdata` package as well.

File: sdk/metric/exemplar/fixed_size_reservoir.go
Comment (dashpole) [reply sdk/metric/exemplar/fixed_size_reservoir.go]: done

File: CHANGELOG.md
Comment (dashpole) [reply CHANGELOG.md]: done

**Review by MrAlias — APPROVED** [review]

**Comment by MrAlias** [comment]
Planned follow up discussed in SIG meeting: Audit the compliance of what was just merged with the specification.

TODO: track ^ in an issue.

**Comment by pellared** [comment]
LGTM. Should we add some testable example showcasing exemplar-reservoir feature?

**Pull request merged**
